scripts/exact_1d_bayesian_filtering_demo.py

```diff
@@ -0,0 +1,515 @@
+        p_pred = p_pred.at[k, :].set(jnp.sum(p_joint, axis=1))
+        p_pred = p_pred.at[k, :].set(p_pred[k, :] / jnp.sum(p_pred[k, :]))
+
+        # p(xk | y(1:k))
+        p_filter = p_filter.at[k, :].set(
+            p_pred[k, :] * y_likelihood(y[k], x_grid, e_pdf, h)
+        )
+        p_filter = p_filter.at[k, :].set(p_filter[k, :] / jnp.sum(p_filter[k, :]))
+
+    p_smooth = jnp.array(p_filter.copy())
+    for k in range(max_iter - 1, -1, -1):
+        # p(xk, xk-1 | y(1:k-1))
+        px = x_pdf(k=k, x_new=X.T, x=X, v_pdf=v_pdf, f=f)
+        px = px * p_smooth[k + 1, :].T / p_pred[k + 1, :].T
+
+        p_smooth = p_smooth.at[k, :].set(jnp.sum(px, axis=1))  # marginalize
+        p_smooth = p_smooth.at[k, :].set(
+            p_smooth[k, :] * p_filter[k, :]
+        )  # multiply p(xk|yk)
+        p_smooth = p_smooth.at[k, :].set(p_smooth[k, :] / jnp.sum(p_smooth[k, :]))
+
+    return p_filter / delta, p_pred / delta, p_smooth / delta
+
+
+def plot_density(
+        x_true, y, inv_h,
+        x_grid, p_pred, p_filter,
+        p_smooth=None, k=1, legend=True,
+        ax=None, vfill=None, title="",
+        linewidth=4.5):
+    if ax is None:
+        fig, ax = plt.subplots(figsize=(12, 8))
+
+    ax.plot(x_grid, p_pred[k], label="Prediction", linewidth=linewidth)
+    ax.plot(x_grid, p_filter[k], label="Filtering", color="k", linewidth=linewidth)
+    if p_smooth is not None:
+        ax.plot(
+            x_grid, p_smooth[k], label="Smoothing", color="orange", linewidth=linewidth
+        )
+
+    y_max = max(p_pred[k].max(), p_filter[k].max()) * 1.05
+    if p_smooth is not None:
+        y_max = max(y_max, p_smooth[k].max()) * 1.05
+    ax.vlines([x_true[k]], ymin=0, ymax=y_max, label="True state", color="k")
+    ax.vlines(
+        inv_h(y[k]),
+        ymin=0,
+        ymax=y_max,
+        color="r",
+        label="Measurement",
+    )
+    if vfill is not None:
+        ax.axvspan(*vfill, color="lightgrey", alpha=0.4, label="Measurement range")
+    ax.set_ylim(0)
+    ax.set_ylabel(f"$p(x_{{{k}}}|y_{{1:{k}}})$")
+    ax.set_xlabel("x")
+    if legend:
+        ax.legend(prop={"size": 16})
+    if title:
+        ax.set_title(title)
+
+
+def plot_densities(
+        x_true, y, inv_h, x_grid, p_pred, p_filter, p_smooth, max_iter, legend=True):
+    nplot = max_iter + 1
+    ncol = 10
+    nrow = nplot // ncol + 1
+    fig, axes = plt.subplots(nrow, ncol, figsize=(12, 6), sharex=True, sharey=True)
+    axes = axes.ravel()
+
+    plt.suptitle("All density plots to look for weird pattern")
+
+    for k in range(1, max_iter + 1):
+        plot_density(
+            x_true, y, inv_h, x_grid=x_grid,
+            p_pred=p_pred, p_filter=p_filter, p_smooth=p_smooth,
+            k=k, ax=axes[k], legend=False,
+            linewidth=1.5,
+        )
+    fig.tight_layout()
+
+
+def experiment_setup(
+    rng_key, grid_minval, grid_maxval,
+    num_grid_points, x0_rvs, v_rvs,
+    e_rvs, f, h,
+    max_iter, plot_xy=False):
```

Comment:
Indent to distinguish from rest of code: see https://peps.python.org/pep-0008/#indentation